### PR TITLE
feat: initial support for cache api

### PR DESCRIPTION
**Pull request by caass**
Add initial support for the cache api.

Notable things:
- There's some size cost in monomorphization of `Into<CacheKey>` params, but I'm not sure that's avoidable without presenting a grosser API with a `String` version and a `Request` version
- The `Response` params have to passed by value since they don't implement `Clone`. Attempting to `#[derive(clone)]` on `Response` errors with `Body has already been used. It can only be used once. Use tee() first if you need to read it twice.`
- `cache.match` was renamed to `cache.get` to avoid conflicting with the `match` keyword.

### CODE REVIEW

File: worker/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            inner: get_default_cache(),
+        }
+    }
+}
+
+impl Cache {
+    pub async fn open<S: Into<String>>(name: S) -> Self {
+        // unwrap is safe because this promise never rejects
+        // https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage/open
+        let inner = JsFuture::from(get_cache_from_name(name.into()))
```
[the remark this answers is not in the document]
Comment (caass) [reply worker/src/cache.rs:27]: good call

File: worker/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            .await
+            .unwrap()
+            .into();
+        Self { inner }
+    }
+
+    pub async fn put(&self, request: &Request, response: Response) -> Result<()> {
+        // TODO take the request by value and get the inner via From. for some reason that isn't working so...here we are
+        // i don't think it's actually too expensive? like it's just cloning a pointer i think.
+        let ffi_request = request.inner().clone()?;
```
[the remark this answers is not in the document]
Comment (caass) [reply worker/src/cache.rs:41]: no but i'm about to find out :)

**Comment by caass** [comment]
I tested this a little bit, `put` and `get` are working:

```rust
use worker::*;

mod utils;

static ORIGIN: &str = "https://example.org/";

#[event(fetch)]
pub async fn main(_req: Request, _env: Env) -> Result<Response> {
    // Optionally, get more helpful error messages written to the console in the case of a panic.
    utils::set_panic_hook();

    let cache = Cache::default();

    if let Some(cached) = cache.get(ORIGIN, true).await? {
        console_log!("Got response from cache!");
        Ok(cached)
    } else {
        let response = Fetch::Url(Url::parse(ORIGIN)?).send().await?;
        cache.put(ORIGIN, response).await?;
        console_log!("Cached response!");
        Response::ok("cached response")
    }
}
```

**Comment by nilslice** [comment]
I was able to get your example working, but any thoughts why this example isn't working? I'd expect to see the same timestamp in subsequent responses... 

```rust
let cache = Cache::default();
let key = req.url()?.to_string();
if let Some(resp) = cache.get(&key, true).await? {
    return Ok(resp);
} else {
    let resp = Response::ok(format!("timestamp: {}", Date::now().as_millis()))?;
    cache.put(&key, &resp).await?;
    return Ok(resp);
}
```

File: worker-sys/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+extern "C" {
+    #[wasm_bindgen(js_name = Cache)]
+    #[derive(Debug, Clone, PartialEq)]
+    pub type Cache;
+
+    #[wasm_bindgen(structural, method, js_class=Cache, js_name = put)]
+    pub fn put_request(this: &Cache, request: Request, response: Response) -> ::js_sys::Promise;
+
+    #[wasm_bindgen(structural, method, js_class=Cache, js_name = put)]
+    pub fn put_url(this: &Cache, url: String, response: Response) -> ::js_sys::Promise;
```
Comment (dbw9580) [inline worker-sys/src/cache.rs:15]: Here and below, can we use `&str` instead of `String`?

**Comment by Lyrex** [comment]
Is there any way to improve the ergonomics of the API? I tried implementing the example from @nilslice, but for me that doesn't work (as stated in the initial PR) because the implemented cache api can't handle response references. I imagine that you'd want to cache the response you're about to send out, so for subsequent requests you can get it from cache. With the current implementation, you have to do some (in my opinion) awkward workarounds like cloning the content and the headers and creating new requests.

This is the example I came up with:
```rust
let cache = Cache::default();
let key = req.url()?.to_string();
if let Some(cache_response) = cache.get(&cache_key, true).await? {
    return Ok(cache_response);
} else {
    let resp = Response::ok(format!("timestamp: {}", Date::now().as_millis()))?;
    let response_content = resp.bytes().await?.clone();
    let response_headers = resp.headers().clone();

   // build response that gets saved to cache
    let cache_response = Response::from_bytes(response_content.clone())
        .unwrap()
        .with_headers(response_headers.clone());
    // save to cache
    let _ = cache.put(&req, cache_response).await;

    // build response that gets sent out to the user
    let response = Response::from_bytes(response_content.clone())
        .unwrap()
        .with_headers(response_headers.clone());

    // send response
    return Ok(response);
}
```

For me, if I don't do all those `.clone()` shenanigans I either get corruped data on the first request (the one that saves to the cache) or on subsequent request (because the cache is trashed). I did my experiments with image data, so that the responses were a few hundred KB in case that makes a difference. 

But other than that, after implementing those workarounds the cache api implementation seems to work fine for me. I implemented some features using `get`, `put` and `delete` and so far it all looks good.

**Comment by nilslice** [comment]
Maybe @caass has an update here?

**Comment by Dav1dde** [comment]
You can actually convert into a `EdgeResponse` (`worker_sys::Response`), clone that and convert back into a `worker::Response`, this has the advantage of also working with streaming responses. `.clone()` is also what the JS docs seem to be using.

This is still very untested, it seems to work in Miniflare though:
```rust
impl ResponseExt for Response {
    fn with_header(mut self, name: &str, value: &str) -> crate::Result<Self> {
        self.headers_mut().set(name, value)?;
        Ok(self)
    }

    fn cloned(self) -> crate::Result<(Self, Self)> {
        let headers = self.headers().clone();

        let response1: worker_sys::Response = self.into();
        let response2 = response1.clone()?;

        let body1 = worker::ResponseBody::Stream(response1);
        let body2 = worker::ResponseBody::Stream(response2);

        let response1 = worker::Response::from_body(body1)?.with_headers(headers.clone());
        let response2 = worker::Response::from_body(body2)?.with_headers(headers);

        Ok((response1, response2))
    }
}
```

Be careful though with streaming responses, they completely ignore modifications on the `worker::Request` object.

**Comment by Dav1dde** [comment]
Unfortunately there is bad news:

```rust
    let cache = Cache::default();
    let use_cache = req.method() == Method::Get;

    if use_cache {
        // TODO: figure out why this doesn't seem to work on workers -> cache always empty
        if let Some(cached) = cache.get(&*req, true).await? {
            log::debug!("cache hit");
            // TODO: 304 handling?
            return cached.with_header("Cf-Cache-Status", "HIT");
        }
    }

    // TODO: figure this lifetime out, the clone may not be required
    let request_for_cache = req.clone()?;

    let response = f(req, env, ctx).await?;

    if use_cache {
        let (response, response_for_cache) = response.cloned()?;

        ctx.wait_until(async move {
            log::debug!("--> caching response");
            let _ = cache.put(&request_for_cache, response_for_cache).await;
            log::debug!("<-- response cached");
        });

        bindgen::Response::persist(response.with_header("Cf-Cache-Status", "MISS")?)
    } else {
        Ok(response)
    }
```

This code does not seem to store anything in the cache (every response is a `MISS`). It makes no difference if I use `waitUntil` or not and it works on Miniflare.

Also the response body seems to be consumed properly, if I replace it with a dummy response I get the unnecessary clone warning:
> Your worker called response.clone(), but did not read the body of both clones. This is wasteful, as it forces the system to buffer the entire response body in memory, rather than streaming it through. This may cause your worker to be unexpectedly terminated for going over the memory limit. If you only meant to copy the response headers and metadata (e.g. in order to be able to modify them), use `new Response(response.body, response)` instead.

I'll try not using the request as a cache key but the url instead.

**Comment by ByteAlex** [comment]
Hi, what's the state on this PR at the moment? Is the person maintaining this PR still active? I may pick it up for them if they're not.

**Comment by aidantwoods** [comment]
Is there anything that needs doing to help get this finished merged? Looks like the separate mangled responses issue was resolved

**Comment by zebp** [comment]
Closing this due to a messy git history. These changes are now in #182

**Pull request closed without merging**